File: auto_verbs/verbs/apartments_com/probe_cards.js
/**
 * Probe: Inspect apartments.com listing card DOM for price/beds selectors.
 * Uses Stagehand (better anti-detection) to navigate via search bar.
 */
const { Stagehand } = require("@browserbasehq/stagehand");
const { setupLLMClient } = require("../../stagehand-utils");
const path = require("path");
const os = require("os");

async function main() {
  const llmClient = setupLLMClient("hybrid");
  const stagehand = new Stagehand({
    env: "LOCAL", verbose: 0, llmClient,
    localBrowserLaunchOptions: {
      userDataDir: path.join(os.homedir(), "AppData", "Local", "Google", "Chrome", "User Data", "Default"),
      headless: false,
      viewport: { width: 1920, height: 1080 },
      args: ["--disable-blink-features=AutomationControlled", "--disable-infobars", "--disable-extensions"],
    },
  });
  await stagehand.init();
  const page = stagehand.context.pages()[0];

  // Navigate via search
  await page.goto("https://www.apartments.com");
  await page.waitForLoadState("domcontentloaded");
  await page.waitForTimeout(5000);

  // Search
  await stagehand.act("Click the search bar to start typing a location");
  await page.waitForTimeout(500);
  await page.type("Austin, TX", { delay: 80 });
  await page.waitForTimeout(2500);
  await stagehand.act("Click the first autocomplete suggestion for Austin, TX");
  await page.waitForLoadState("domcontentloaded");
  await page.waitForTimeout(5000);
  console.log("URL:", await page.evaluate("window.location.href"));

  // Inspect first 3 cards
  const cardInfo = await page.evaluate(`(() => {
    const cards = document.querySelectorAll('article.placard, [data-listingid]');
    const results = [];
    for (let i = 0; i < Math.min(3, cards.length); i++) {
      const card = cards[i];
      // Get all child elements with their classes and text
      const children = [];
      card.querySelectorAll('*').forEach(el => {
        const cls = el.className && typeof el.className === 'string' ? el.className : '';
        const text = el.textContent.trim().substring(0, 100);
        const tag = el.tagName.toLowerCase();
        if (text && (cls.includes('price') || cls.includes('pricing') || cls.includes('rent') ||
            cls.includes('bed') || cls.includes('bath') || cls.includes('unit') ||
            cls.includes('property-') || tag === 'p' || tag === 'span')) {
          children.push({ tag, cls: cls.substring(0, 120), text });
        }
      });
      // Also get all <p> and <span> children directly
      const pSpans = [];
      card.querySelectorAll('p, span, div.price-range, div.property-pricing').forEach(el => {
        const cls = el.className && typeof el.className === 'string' ? el.className : '';
        const text = el.textContent.trim().substring(0, 100);
        if (text) pSpans.push({ tag: el.tagName.toLowerCase(), cls: cls.substring(0, 120), text });
      });
      results.push({
        outerHTMLSnippet: card.outerHTML.substring(0, 1500),
        innerText: card.textContent.trim().substring(0, 300),
        matchingChildren: children.slice(0, 20),
        allPSpans: pSpans.slice(0, 15),
      });
    }
    return { totalCards: cards.length, cards: results };
  })()`);

  console.log("Total cards:", cardInfo.totalCards);
  for (let i = 0; i < cardInfo.cards.length; i++) {
    console.log(`\n${"=".repeat(60)}\nCARD ${i}:`);
    console.log("--- Inner text ---");
    console.log(cardInfo.cards[i].innerText);
    console.log("\n--- Matching children (price/bed/rent/property-*) ---");
    cardInfo.cards[i].matchingChildren.forEach(c => {
      console.log(`  <${c.tag}> class="${c.cls}" → "${c.text}"`);
    });
    console.log("\n--- All <p>/<span> ---");
    cardInfo.cards[i].allPSpans.forEach(c => {
      console.log(`  <${c.tag}> class="${c.cls}" → "${c.text}"`);
    });
    console.log("\n--- HTML snippet ---");
    console.log(cardInfo.cards[i].outerHTMLSnippet.substring(0, 800));
  }

  await stagehand.close();
  console.log("\nDone!");
}
main().catch(e => { console.error(e.message); process.exit(1); });

File: auto_verbs/verbs/apartments_com/probe_cards.py
"""Probe apartments.com listing card DOM to discover correct selectors for price/beds."""
import os
from playwright.sync_api import sync_playwright

def main():
    user_data_dir = os.path.join(
        os.environ["USERPROFILE"],
        "AppData", "Local", "Google", "Chrome", "User Data", "Default",
    )
    with sync_playwright() as pw:
        ctx = pw.chromium.launch_persistent_context(
            user_data_dir, channel="chrome", headless=False,
            viewport={"width": 1920, "height": 1080},
            args=["--disable-blink-features=AutomationControlled", "--disable-infobars", "--disable-extensions"],
        )
        page = ctx.pages[0] if ctx.pages else ctx.new_page()

        # Go directly to a results page with price filter
        page.goto("https://www.apartments.com/austin-tx/1000-to-2000/")
        page.wait_for_load_state("domcontentloaded")
        page.wait_for_timeout(5000)
        print(f"URL: {page.url}")
        print(f"Title: {page.title()}")

        # Inspect first 3 article.placard cards
        cards = page.locator("article.placard")
        count = cards.count()
        print(f"\nFound {count} article.placard cards")

        for i in range(min(3, count)):
            card = cards.nth(i)
            print(f"\n{'='*60}")
            print(f"CARD {i}:")
            # Get the outer HTML (first 2000 chars)
            html = card.evaluate("el => el.outerHTML")
            print(html[:2000])
            print(f"\n--- Inner text ---")
            print(card.inner_text()[:500])

            # Try various selectors for price
            print(f"\n--- Price selectors ---")
            for sel in [
                '[class*="property-pricing"]', 'p.property-pricing',
                'span[class*="price"]', 'div[class*="pricing"]',
                '[class*="rent"]', '[class*="Price"]',
                'p.price-range', '.price-range',
                'span.property-rents', 'div.price-range',
            ]:
                try:
                    el = card.locator(sel).first
                    if el.count() > 0:
                        txt = el.inner_text(timeout=1000).strip()
                        print(f"  {sel} → \"{txt}\"")
                except Exception:
                    pass

            # Try various selectors for beds/baths
            print(f"\n--- Beds selectors ---")
            for sel in [
                '[class*="property-beds"]', 'p.property-beds',
                'span[class*="bed"]', '[class*="beds"]',
                '[class*="Bed"]', '[class*="unit-type"]',
            ]:
                try:
                    el = card.locator(sel).first
                    if el.count() > 0:
                        txt = el.inner_text(timeout=1000).strip()
                        print(f"  {sel} → \"{txt}\"")
                except Exception:
                    pass

        # Also dump all unique class names inside first card
        if count > 0:
            print(f"\n{'='*60}")
            print("ALL CLASS NAMES in first card:")
            classes = cards.nth(0).evaluate("""el => {
                const all = el.querySelectorAll('*');
                const classSet = new Set();
                all.forEach(e => {
                    if (e.className && typeof e.className === 'string') {
                        e.className.split(/\\s+/).forEach(c => { if (c) classSet.add(c); });
                    }
                });
                return Array.from(classSet).sort();
            }""")
            print(", ".join(classes))

        ctx.close()
        print("\nDone!")

if __name__ == "__main__":
    main()
